accounting/invoice/tests/test_tasks.py

```diff
@@ -0,0 +1,190 @@
+
+    @freezegun.freeze_time(NOW)
+    def setUp(self):
+        """ Set up test objects. """
+        self.create_client_and_provider_links()
+
+        # Create invoices for the month. The task will consider these the final ones.
+        self.invoice1 = factories.InvoiceFactory(date=NOW, provider=self.provider1, client=self.client)
+        self.invoice2 = factories.InvoiceFactory(date=NOW, provider=self.provider2, client=self.client)
+        super().setUp()
+
+    @mock.patch.object(models.Invoice, 'upload_to_google_drive')
+    @mock.patch.object(models.Invoice, 'fill_line_items_from_jira')
+    @mock.patch.object(EmailMessage, 'attach_file')
+    @mock.patch.object(models.Invoice, 'to_pdf')
+    def test_send_final_invoices(self, mock_to_pdf, mock_attach_file, mock_fill_line_items_from_jira,
+                                 mock_upload_to_google_drive):
+        """ The invoice finalization email sent out every month contains proper information. """
+        # Mock.
+        pdf_path = 'https://drive.google.com/invoice.pdf'
+        mock_upload_to_google_drive.return_value = {'webContentLink': pdf_path}
+        mock_to_pdf.return_value = pdf_path
+
+        # Call.
+        tasks.send_final_invoices()
+        self.invoice1.refresh_from_db()
+        self.invoice2.refresh_from_db()
+
+        # The invoices will be marked as automatically approved since nobody visited the URL.
+        self.assertTrue(self.invoice1.is_approved)
+        self.assertTrue(self.invoice2.is_approved)
+        self.assertEqual(self.invoice1.approved, choices.InvoiceApproval.automatically)
+        self.assertEqual(self.invoice2.approved, choices.InvoiceApproval.automatically)
+
+        # 2 providers, so 2 of everything.
+        self.assertEqual(models.Invoice.objects.all().count(), 2)
+        self.assertEqual(mock_to_pdf.call_count, 2)
+        self.assertEqual(mock_attach_file.call_count, 2)
+        mock_fill_line_items_from_jira.assert_not_called()
+        self.assertEqual(mock_upload_to_google_drive.call_count, 2)
+        self.assertEqual(len(mail.outbox), 2)
+
+        # Email subjects should be the same.
+        self.assertEqual(mail.outbox[0].subject, mail.outbox[1].subject)
+        self.assertEqual(mail.outbox[0].subject, 'Your approved invoice for December')
+
+        # Body contains proper invoice approval URL and text, for both providers.
+        body = (
+            "Hello,\n"
+            "\n"
+            "I have attached your finalized and approved invoice for December.\n"
+            "\n"
+            "We are now proceeding to send the bank transfers, and you can expect payment to arrive "
+            "in your bank account within 5-8 business days at most.\n"
+            "\n"
+            "If you have any issues, please contact billing@opencraft.com.\n"
+        )
+        self.assertEqual(mail.outbox[0].body, body.format(self.invoice1.uuid))
+        self.assertEqual(mail.outbox[1].body, body.format(self.invoice2.uuid))
```

Comment:
Yeap I'm just trying to get this out the way given the time I have for such a large feature surface. I could always break tasks down and have just done a single function in `invoice.tasks` per week, allowing myself plenty of time to refactor this quite well, but that wouldn't suffice our requirements.

Reply:
@UmanShahzad I actually think the existing code reads really well. It can be nice and/or helpful for test code to be a bit more verbose. And sometimes figuring out where to draw the line isn't straightforward. So just in case my previous comment came across as though I was complaining: I wasn't :)

Reply:
Haha, yep no worries :+1: